### DISCUSSION TITLE
Audio redirection won't work if PulseAudio is started by per-user systemd

### DISCUSSION THREAD

**Discussion by metalefty** [neutrinolabs/xrdp#1910] [Ideas]
xrdp uses the following environment variables to pass the locations of audio socket files and socket files directory between xrdp session and PulseAudio module. See also https://github.com/neutrinolabs/pulseaudio-module-xrdp/search?q=getenv&type=code.

* XRDP_SOCKET_PATH
* XRDP_PULSE_SINK_SOCKET
* XRDP_PULSE_SOURCE_SOCKET
* PULSE_SCRIPT

However, recent GNU/Linux systems use systemd to start the per-user PulseAudio process. Not a child process of `xrdp-sesman`. For example, `/etc/systemd/user/default.target.wants/pulseaudio.service ` is the target to start PulseAudio process.

This is an example of the environment variables `pulseaudio` process started from systems can see. Necessary environment variables are not visible from systems `pulseaudio`. As a result, unfortunately, audio redirection does not work.

```
DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus
HOME=/home/meta
LANG=ja_JP.UTF-8
LOGNAME=meta
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin
SHELL=/bin/bash
USER=meta
XDG_DATA_DIRS=/home/meta/.local/share/flatpak/exports/share:/var/lib/flatpak/exports/share:/usr/local/share:/usr/share
XDG_RUNTIME_DIR=/run/user/1000
NOTIFY_SOCKET=/run/user/1000/systemd/notify
MANAGERPID=115733
LISTEN_PID=115755
LISTEN_FDS=1
LISTEN_FDNAMES=pulseaudio.socket
INVOCATION_ID=85269c9159824369b61b068cb1cb4dab
JOURNAL_STREAM=9:1508197
```

We need to consider how to pass some necessary information from xrdp to the PulseAudio module. 

This issue is originally reported by @CyberTrust and @haw.

**Comment by matt335672** [comment]
I had a brief [Gitter conversation](https://gitter.im/neutrinolabs/xrdp-questions?at=5f7c8d9feb6dbd77f9041be4) about this with @derekschrock back in October last year. At the time we managed to get something working by running the xrdp session with a separate dbus instance. At the time this wasn't possible with Gnome due to its heavy dependence on `systemd --user`, and I suspect over time it will become harder for other desktop environments too.

Something I was playing with was making the environment variables to the modules parameters instead. Then the PULSE_SCRIPT can use `pactl` to load the modules dynamically into the `systemd --user` instance of pulseaudio. I only had a brief play with it, and didn't get it working. However, the code is here:-

https://github.com/matt335672/pulseaudio-module-xrdp/tree/module_params

I've not considered unloading the modules when the session ends yet, or anything like that.

**Comment by abma** [comment]
i've created a login script which loads dynamicly the xrdp-pulse modules:

```
  pactl load-module module-xrdp-sink
  pactl load-module module-xrdp-source
```

this way the systemd-stuff doesn't need to be touched and sound works inside the xrdp session!

**Comment by matt335672** [comment]
Hi @abma 

That's really interesting - thanks.

The modules use environment variables to locate the chansrv endpoints. If these are readable by the pulse audio server, this should work.

In your instance, what is the output of `systemctl --user show-environment`? Did you need to use any `systemctl --user set-environment` variables to get this working?

**Comment by abma** [comment]
```
$ systemctl --user show-environment
[...]
XRDP_PULSE_SINK_SOCKET=xrdp_chansrv_audio_out_socket_11
XRDP_PULSE_SOURCE_SOCKET=xrdp_chansrv_audio_in_socket_11
XRDP_SESSION=1
XRDP_SOCKET_PATH=/run/xrdp/sockdir
```
> Did you need to use any systemctl --user set-environment variables to get this working?
no!

i basicly just created the file /etc/X11/Xsession.d/99xrdp-pulse:
```
pactl load-module module-xrdp-sink
pactl load-module module-xrdp-source
```

(its sourced, no +x / shebang required)

**Reply by matt335672** [reply]
Thanks for the additional info.

On Ubuntu, the following command in `/etc/X11/Xsession.d/95dbus_update-activation-env` seems to be doing all the magic for us:-

```
dbus-update-activation-environment --verbose --systemd --all
```

Provided your login script runs after this line, you should be fine.

Currently xrdp is set up to provide the environment "PULSE_SCRIPT" for use by the PulseAudio daemon. That method doesn't work for systemd-based systems, as pulseaudio is started when the PAM session is started (which is before the windows session runs).

I created the following script on my system under `/etc/X11/Xsession.d/90xrdp-pulse`. Permissions are 644.

```sh

# This file is sourced by Xsession(5), not executed.

if [ -n "$XRDP_SESSION" -a \
     -n "$DISPLAY" -a \
     -n "$XRDP_SOCKET_PATH" -a \
     -n "$XRDP_PULSE_SINK_SOCKET" -a \
     -n "$XRDP_PULSE_SOURCE_SOCKET" ]
then
     if systemctl --user is-active pulseaudio >/dev/null 2>&1
     then
         systemctl --user set-environment \
             DISPLAY="$DISPLAY" \
             XRDP_SOCKET_PATH="$XRDP_SOCKET_PATH" \
             XRDP_PULSE_SINK_SOCKET="$XRDP_PULSE_SINK_SOCKET" \
             XRDP_PULSE_SOURCE_SOCKET="$XRDP_PULSE_SOURCE_SOCKET"
        pactl load-module module-xrdp-sink
        pactl load-module module-xrdp-source
    fi
fi
```

This doesn't depend on `dbus-update-activation-environment` being called, and also won't try to load xrdp drivers for non-xrdp sessions. It seems to work on Ubuntu 20.04, apart from some latency issues which I haven't looked at yet.

Any chance you could try that on your system? If it's OK there, we can add it to the wiki.

**Comment by abma** [comment]
sadly the "pactl load-module" way didn't work stable for me. sometimes i had sound, sometimes not. the pulseaudio service and/or process sometimes seems to stay running after logout.

i've reported it upstream, maybe they have an idea: https://gitlab.freedesktop.org/pulseaudio/pulseaudio/-/issues/1271

**Reply by matt335672** [reply]
Well, thanks for looking at at it and reporting back.

**Comment by matt335672** [comment]
@abma - please take a look at neutrinolabs/pulseaudio-module-xrdp#59 which will (I think) sort everything out for you. Feel free to build it from my repo directly:-

```
git clone https://github.com/matt335672/pulseaudio-module-xrdp.git
cd pulseaudio-module-xrdp
git checkout module_params
```

**Reply by abma** [reply]
in a quick test i noticed it seems more reliable: i have sound instantly, without the patch it needed ~10 seconds until sound started.

for now i just updated the modules: when it works nice (on a server with ~50 users) i will try again starting it via systemd.

thanks!

**Comment by matt335672** [comment]
pulseaudio-module-xrdp 0.6 has been released which now supports pulseaudio being started by per-user systemd.
